src/blockassembler.cpp

```diff
@@ -0,0 +1,492 @@
+    pblock->nNonce         = 0;
+    pblocktemplate->vTxSigOps[0] = GetLegacySigOpCount(*(pblock->vtx[0]));
+    appendSaplingTreeRoot();
+
+    if (fProofOfStake) { // this is only for PoS because the IncrementExtraNonce does it for PoW
+        pblock->hashMerkleRoot = BlockMerkleRoot(*pblock);
+        LogPrintf("CPUMiner : proof-of-stake block found %s \n", pblock->GetHash().GetHex());
+        if (!SignBlock(*pblock, *pwallet)) {
+            LogPrintf("%s: Signing new block with UTXO key failed \n", __func__);
+            return nullptr;
+        }
+    }
+
+    CValidationState state;
+    if (!TestBlockValidity(state, *pblock, pindexPrev, false, false)) {
+        throw std::runtime_error(strprintf("%s: TestBlockValidity failed: %s", __func__, FormatStateMessage(state)));
+    }
+
+    return std::move(pblocktemplate);
+}
+
+bool BlockAssembler::isStillDependent(CTxMemPool::txiter iter)
+{
+    for (CTxMemPool::txiter parent : mempool.GetMemPoolParents(iter)) {
+        if (!inBlock.count(parent)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool BlockAssembler::TestForBlock(CTxMemPool::txiter iter)
+{
+    unsigned int nTxSize = iter->GetTxSize();
+    if (nBlockSize + nTxSize >= nBlockMaxSize) {
+        // If the block is so close to full that no more txs will fit
+        // or if we've tried more than 50 times to fill remaining space
+        // then flag that the block is finished
+        if (nBlockSize >  nBlockMaxSize - 100 || lastFewTxs > 50) {
+            blockFinished = true;
+            return false;
+        }
+        // Once we're within 1000 bytes of a full block, only look at 50 more txs
+        // to try to fill the remaining space.
+        if (nBlockSize > nBlockMaxSize - 1000) {
+            lastFewTxs++;
+        }
+        return false;
+    }
+
+    const bool isShielded = iter->IsShielded();
+    if (isShielded && nSizeShielded + nTxSize > MAX_BLOCK_SHIELDED_TXES_SIZE) {
+        return false;
+    }
+
+    if (nBlockSigOps + iter->GetSigOpCount() >= MAX_BLOCK_SIGOPS_CURRENT) {
+        // If the block has room for no more sig ops then
+        // flag that the block is finished
+        if (nBlockSigOps > MAX_BLOCK_SIGOPS_CURRENT - 2) {
+            blockFinished = true;
+            return false;
+        }
+        // Otherwise attempt to find another tx with fewer sigops
+        // to put in the block.
+        return false;
+    }
+
+    // Legacy zerocoin transactions are disabled forever.
+    assert(!iter->GetSharedTx()->ContainsZerocoins());
+
+    // Must check that lock times are still valid
+    // This can be removed once MTP is always enforced
+    // as long as reorgs keep the mempool consistent.
+    if (!IsFinalTx(iter->GetTx(), nHeight))
+        return false;
+
+    // Don't add shielded transactions if Sapling isn't active
+    if(sporkManager.IsSporkActive(SPORK_20_SAPLING_MAINTENANCE) && isShielded) {
+        return false;
+    }
```

Comment:
nit: might be good to move this check (as well as the zerocoin txes assertion) at the beginning of the function (to exit/throw early in case).